mcp23xxx: make devicetree compatibles unique per chip

arbrauns opened the pull request:
Pin definitions should correctly reflect the actual drive mode of the GPIO controller, either push-pull or open drain.

EDIT: see https://github.com/zephyrproject-rtos/zephyr/pull/65797#issuecomment-1993904939 for current description of the changes

Fixes #56280.

arbrauns committed:
drivers: mcp23xxx: add support for open-drain chip variants

Pin definitions should correctly reflect the actual drive mode of the GPIO
controller, either push-pull or open drain.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

henrikbrixandersen commented on dts/bindings/gpio/microchip,mcp23xxx.yaml:
```diff
@@ -31,6 +31,12 @@ properties:
     description: |
       Number of gpios supported by the chip.
 
+  open-drain:
+    type: boolean
+    description: |
+      Specifies that the chip has open-drain rather than push-pull outputs. Should be set
+      for mcp23x09 and mcp23x18 chips.
```
Sounds like we ought to have separate bindings for these chips instead?

arbrauns replied:
Maybe, but I haven't found a nice way to do that without needlessly copying a bunch of code/configuration when the only difference is really just a single bit.

arbrauns committed:
drivers: remove legacy mcp23s17 driver

This chip is handled by the more generic mcp23xxx driver, which will get a
microchip,mcp23s17 compatible binding in the next commit.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

arbrauns committed:
drivers: add bindings for all existing mcp23xxx variants

This allows getting rid of the ngpios property, which is implicit in the
part number. It also prepares for configuring pins as open-drain on
supporting chips in the next commit.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

arbrauns commented:
I've made quite a few changes now:

1. The legacy `mcp23s17` driver has been removed, since its functionality is covered by `mcp23xxx`
2. Devicetree bindings have been changed to exact chip types (e.g. `microchip,mcp23s09` instead of `microchip,mcp23sxx`)
3. The number of GPIOs and the output type (OD/PP) is now automatically inferred from the chip type, removing the need for the `ngpio` devicetree property

This is a breaking change (users of the chips have to change their devicetree compatibles and remove the obsolete properties), but I think it's more convenient overall.

arbrauns commented:
@jvasanth1 could I get a review?

jvasanth1 commented:
@con-pax could you help to review?

con-pax commented:
> @jvasanth1 could I get a review?

Hey, apologies for the delay. Will do

jvasanth1 commented:
Hi @MaureenHelm, looks like Microchip is not the code owners for the mcp23xxx gpio controller drivers

MaureenHelm commented:
> Hi @MaureenHelm, looks like Microchip is not the code owners for the mcp23xxx gpio controller drivers

How so? The zephyrbot assigned this PR to you based on the MAINTAINERS file, so it looks like you're the code owner to me.

jvasanth1 commented:
> > Hi @MaureenHelm, looks like Microchip is not the code owners for the mcp23xxx gpio controller drivers
> 
> How so? The zephyrbot assigned this PR to you based on the MAINTAINERS file, so it looks like you're the code owner to me.

I think the MAINTAINERS file needs to be updated.

arbrauns commented:
@con-pax ping

henrikbrixandersen reviewed (changes requested):
Please add entries to the migration guide for the removed devicetree compatibles and properties.

aescolar commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -335,6 +335,14 @@ Flash
+  The ``ngpios`` devicetree property has been removed, since it is implied by the model name.
+  Chip variants with open-drain outputs (``mcp23x09``, ``mcp23x18``) now correctly reflect this in
+  their driver API, users of these devices should ensure they pass appropriate values to
+  :c:func:`gpio_pin_set`.
```
```suggestion
  :c:func:`gpio_pin_set` (:github:`65797`).
```

decsny commented:
The explosion of DT bindings on various PRs because people are making a new file for every single part number and sub part number is really bothering me. We will have to address this soon with a more refined solution than having a gazillion files

arbrauns commented:
Sure, some form of templating for binding files would be nice. Having more precise compatible names for devices is good though, imo.

decsny commented:
> Sure, some form of templating for binding files would be nice. Having more precise compatible names for devices is good though, imo.

yes, we should have the specific part number as the compatible string, that is according to the DT spec, the issue is the zephyr infrastructure around this is not scaling to meet that requirement

arbrauns commented:
Ping, seems like both @jvasanth1 and @con-pax are unavailable.

arbrauns commented:
@henrikbrixandersen is there any way I can roll for a new assignee?

henrikbrixandersen reviewed (changes requested):
Looks good, only one minor nit.

henrikbrixandersen commented:
Reassigning to myself due to lack of activity from @jvasanth1.

@arbrauns Please rebase on `main` and resolve the merge conflict.

arbrauns commented on dts/bindings/gpio/microchip,mcp23s17.yaml:
```diff
@@ -1,26 +1,11 @@
 #
-# Copyright (c) 2020 Geanix ApS
+# Copyright (c) 2024 SILA Embedded Solutions GmbH
```
If you check the commits, I completely removed that file and then replaced it with new content that is unrelated to the previous content, therefore the old copyright doesn't apply to it.

sknsean commented:
LGTM 👍
